decidim-bulletin_board-ruby/lib/decidim/bulletin_board/client.rb

```diff
@@ -34,13 +34,18 @@ def configured?
         private_key && server && api_key
       end
 
+      def sign_data(data)
+        JWT.encode(data, identification_private_key.keypair, "RS256")
+      end
+
       def setup_election(election_data)
         message_id = "#{election_data[:election_id]}.create_election+a.#{authority_slug}"
         Decidim::BulletinBoard::CreateElection.call(election_data, message_id)
       end
 
-      def cast_vote
-        cast_vote = Decidim::BulletinBoard::Voter::CastVote.new
+      def cast_vote(election_data, voter_data, encrypted_vote)
+        form = Decidim::BulletinBoard::Voter::VoteForm.new(self, election_data, voter_data, encrypted_vote)
```

Comment:
Did you think about exposing the form to the app and receiving it as an argument? That way you would be able to validate the data before calling the method, returning error messages and would be a good approach for the `create_election` method. What do you think?

Reply:
I really don't know 😄! I don't like `decidim` knowing too much about implementation details. I mean with the current implementation decidim just needs to know this:

```ruby
begin
  bulletin_board_client.cast_vote(...)
rescue
  # something bad happens (i.e. the provided data is wrong)
end
```

That means it doesn't need to know about forms or commands or any internal stuff, and we can change the gem implementation when we want without changing the public API.

But I also agree that using just `Hash` could be very error prone, so I am not 100% sure about the previous statement. I agree that we should refactor the `create_election` method to hide the `election_data` implementation details, but I am not sure how to do it at the moment.

When we do that refactor we can discuss the different approaches if you want. Sounds good? 😄